File: 1408-string-matching-in-an-array/1408-string-matching-in-an-array.java
class Solution {
    public List<String> stringMatching(String[] words) {
        // build trie for each permutation of each string
        var root = new TrieNode();
        for(var word : words)
            for(var start = 0; start < word.length(); start++)
                insertWord(root, word.substring(start));
        
        // find substrings
        var matching = new ArrayList<String>();
        for(var word: words)
            if(isSubstring(root, word))
                matching.add(word);
        
        return matching;
    }

    private void insertWord(TrieNode root, String word){
        var curr = root;
        for(var ch : word.toCharArray()){
            if(curr.children[ch-'a'] == null)
                curr.children[ch-'a'] = new TrieNode();
            curr = curr.children[ch-'a'];
            curr.frequency++;
        }
    }

    private boolean isSubstring(TrieNode root, String word){
        var curr = root;
        for(var ch : word.toCharArray())
            curr = curr.children[ch-'a'];

        return curr.frequency > 1;  // freq = 1 is same word
    }
}

class TrieNode {
    int frequency;
    TrieNode[] children;

    TrieNode() {
        frequency = 0;
        children = new TrieNode[26];
    }
}

File: 1408-string-matching-in-an-array/1408-string-matching-in-an-array.cs
public class Solution {
    public IList<string> StringMatching(string[] words) {
        // build trie for each permutation of each string
        var root = new TrieNode();
        foreach(var word in words)
            for(var start = 0; start < word.Length; start++)
                InsertWord(root, word.Substring(start));
        
        // find substrings
        var matching = new List<string>();
        foreach(var word in words)
            if(IsSubstring(root, word))
                matching.Add(word);
        
        return matching;
    }

    private void InsertWord(TrieNode root, string word){
        var curr = root;
        foreach(var ch in word){
            if(curr.children[ch-'a'] == null)
                curr.children[ch-'a'] = new TrieNode();
            curr = curr.children[ch-'a'];
            curr.frequency++;
        }
    }

    private bool IsSubstring(TrieNode root, string word){
        var curr = root;
        foreach(var ch in word)
            curr = curr.children[ch-'a'];

        return curr.frequency > 1;  // freq = 1 is same word
    }
}

public class TrieNode {
    public int frequency;
    public TrieNode[] children;

    public TrieNode() {
        frequency = 0;
        children = new TrieNode[26];
    }
}